### PR TITLE
DAOS-11987 test: Support of SCM tmpfs hugepages

### PR DESCRIPTION
**Pull request by knard-intel**
# Description

When HugePages are enabled with tmpfs, the usable storage size lost when created and destroying a pool is not the same.

More details on how the allowed size is computed could be found in the comments of the tickets [DAOS-11987](https://daosio.atlassian.net/browse/DAOS-11987) and [DAOS-12428](https://daosio.atlassian.net/browse/DAOS-12428)

The pool_create_all_vm was also heavily tested with the CI: 100 iteration of the test has been done with the Jenkins build [PR-11267#2](https://build.hpdd.intel.com/blue/organizations/jenkins/daos-stack%2Fdaos/detail/PR-11267/2/tests)

Use of `Test-tag*` to add weekly tests impacted by this PR.

# Before requesting gatekeeper:

* [x] Two review approvals and any prior change requests have been resolved.
* [x] Testing is complete and all tests passed or there is a reason documented in the PR why it should be force landed and forced-landing tag is set.
* [x] `Features:` (or `Test-tag*`) commit pragma was used or there is a reason documented that there are no appropriate tags for this PR.
* [x] Commit messages follows the guidelines outlined [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments).
* [x] Any tests skipped by the ticket being addressed have been run and passed in the PR.

# Gatekeeper:

* [ ] You are the appropriate gatekeeper to be landing the patch.
* [ ] The PR has 2 reviews by people familiar with the code, including appropriate watchers.
* [ ] Githooks were used. If not, request that user install them and check copyright dates.
* [ ] Checkpatch issues are resolved.  Pay particular attention to ones that will show up on future PRs.
* [ ] All builds have passed.  Check non-required builds for any new compiler warnings.
* [ ] Sufficient testing is done. Check feature pragmas and test tags and that tests skipped for the ticket are run and now pass with the changes.
* [ ] If applicable, the PR has addressed any potential version compatibility issues.
* [ ] Check the target branch.   If it is master branch, should the PR go to a feature branch?  If it is a release branch, does it have merge approval in the JIRA ticket.
* [ ] Extra checks if forced landing is requested
  * [ ] Review comments are sufficiently resolved, particularly by prior reviewers that requested changes.
  * [ ] No new NLT or valgrind warnings.  Check the classic view.
  * [ ] Quick-build or Quick-functional is not used.
* [ ] Fix the commit message upon landing. Check the standard [here](https://daosio.atlassian.net/wiki/spaces/DC/pages/11133911069/Commit+Comments). Edit it to create a single commit. If necessary, ask submitter for a new summary.

### CODE REVIEW

File: src/tests/ftest/pool/create_all_vm.yaml [hunk 1]
```diff
@@ -1,10 +1,12 @@
+setup:
+  start_servers_once: False
 hosts:
   test_servers: 5
 timeouts:
   test_one_pool: 240
   test_rank_filter: 240
   test_two_pools: 160
-  test_recycle_pools: 320
+  test_recycle_pools: 1280
```
Comment (mjmac) [inline src/tests/ftest/pool/create_all_vm.yaml:9]: Why was this large timeout bump needed? That seems concerning...

File: utils/ansible/ftest/templates/daos-make.sh.j2 [hunk 2]
```diff
@@ -197,7 +197,7 @@ if [[ ! -d "$DAOS_INSTALL_DIR" ]] ; then
 fi
 
 cd "$DAOS_SOURCE_DIR"
-SCONS_OPTS="-j $JOBS_NB"
+SCONS_OPTS="-j $JOBS_NB --implicit-deps-changed"
```
Comment (mjmac) [inline utils/ansible/ftest/templates/daos-make.sh.j2:200]: Not sure this was meant to be included in the commit, seems unrelated?

File: src/tests/ftest/pool/create_all_vm.py [hunk 3]
```diff
@@ -144,30 +166,36 @@ def test_recycle_pools(self):
         """
         self.log.info("Test pool creation and destruction")
 
-        for index in range(10):
-            self.log.info("Creating pool %d", index)
+        pool_count = self.params.get("pool_count", "/run/test_recycle_pools/*", 20)
+        delta_bytes = self.get_recycle_pools_delta_bytes(pool_count)
+        scm_avail_bytes = self.get_available_bytes()
+        self.log.debug(
+            "{pool_count} pools to create/destrory with SCM storage"
```
Comment (mjmac) [inline src/tests/ftest/pool/create_all_vm.py:173]: typo: destrory

File: src/tests/ftest/pool/create_all_vm.py [hunk 4]
```diff
@@ -129,6 +129,28 @@ def test_rank_filter(self):
             abs(scm_unused_bytes - self.scm_avail_bytes), delta_bytes,
             msg.format(scm_unused_bytes, self.scm_avail_bytes, delta_bytes))
 
+    def get_recycle_pools_delta_bytes(self, pool_count):
```
Comment (shimizukko) [inline src/tests/ftest/pool/create_all_vm.py:132]: You might want to add description for `pool_count`. Or you could move line 169 inside this method.

**Review by shimizukko — DISMISSED** [review]

File: src/tests/ftest/pool/create_all_vm.yaml (hunk 1 above)
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.yaml:9]: Because I have increase by two the number of iteration (from 10 to 20) and we are now testing with two use cases: hugepage enabled or not. Thus, it is why I increase the timeout of 4.

File: src/tests/ftest/pool/create_all_vm.yaml [hunk 5]
```diff
@@ -17,6 +19,12 @@ server_config:
           class: ram
           scm_mount: /mnt/daos
           scm_size: 4
-          scm_hugepages_disabled: true
+          scm_hugepages_cfg: !mux
+            enabled:
+              scm_hugepages_disabled: false
+            disabled:
+              scm_hugepages_disabled: true
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.yaml:26]: Yep, it will be run one time with huge pages enabled and one time with huge pages disabled.

File: utils/ansible/ftest/templates/daos-make.sh.j2 (hunk 2 above)
Comment (knard-intel) [reply utils/ansible/ftest/templates/daos-make.sh.j2:200]: Yes and No, I needed to properly run ftest on my little cluster. Not sure, it deserves a dedicated PR.

File: src/tests/ftest/pool/create_all_vm.py (hunk 3 above)
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.py:173]: - [ ] Fix typo

File: src/tests/ftest/pool/create_all_vm.py (hunk 4 above)
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.py:132]: - [ ] Add method description

File: src/tests/ftest/pool/create_all_vm.py [hunk 6]
```diff
@@ -129,6 +129,28 @@ def test_rank_filter(self):
...
+        """ Return the allowed size growth of metadata """
+
+        scm_hugepages_disabled = self.params.get(
+            "scm_hugepages_disabled",
+            "/run/server_config/engines/0/storage/0/*",
+            False)
+
+        # These value are empirical ones: management of RAM page fs by the Linux kernel is not
+        # easily predictable.   More details could be found in the comments of the JIRA ticket
+        # DAOS-11987.
```
[the remark this answers is not in the document]
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.py:142]: delta_rank_bytes is the maximal amount of storage lost for all ranks during all the iteration.
Indeed, this last one is increasing in a different way when huge pages are enabled or not: big steps with huge pages vs small steps without huge pages.

- [ ] Add description of the issue of the storage leak issue.

**Comment by knard-intel** [comment]
The flake8 code quality tool is failing on part of the code not updated by this PR.

File: src/tests/ftest/pool/create_all_vm.yaml (hunk 1 above)
Comment (daltonbohning) [reply src/tests/ftest/pool/create_all_vm.yaml:9]: Using `!mux` will run each variant (hugepage enabled or not) as a separate test case. Looking here:
https://build.hpdd.intel.com/job/daos-stack/job/daos/job/PR-11267/5/artifact/Functional%20on%20EL%208/pool/create_all_vm.py/results.html
Each variant took only `430.45s` and `670.13s`

**Comment by daltonbohning** [comment]
> The flake8 code quality tool is failing on part of the code not updated by this PR.

FYI - #11303 to fix those

File: src/tests/ftest/pool/create_all_vm.py [hunk 7]
```diff
@@ -130,16 +137,28 @@ def test_rank_filter(self):
             msg.format(scm_unused_bytes, self.scm_avail_bytes, delta_bytes))
 
     def get_recycle_pools_delta_bytes(self, pool_count):
-        """ Return the allowed size growth of metadata """
-
+        """Return the allowed size of SCM storage space lost for a given number of pools.
+
+        As indicated in JIRA tickets DAOS-11987 and DAOS-12428, some SCM storage are lost when
+        a pool is successively created and destroyed.  This was observed for SCM on RAM and it will
+        be investigated if the same issue arise with SCM on DPCM.  The space lost with SCM on RAM is
```
Comment (shimizukko) [inline src/tests/ftest/pool/create_all_vm.py:144]:
```suggestion
        be investigated if the same issue arise with SCM on DCPM.  The space lost with SCM on RAM is
```

**Review by shimizukko — APPROVED** [review]

File: src/tests/ftest/pool/create_all_vm.yaml (hunk 1 above)
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.yaml:9]: Thanks for the tips, I will adjust the timeout according to the time that you gave me.

- [ ] Adjust timeout

**Comment by knard-intel** [comment]
> > The flake8 code quality tool is failing on part of the code not updated by this PR.
> 
> FYI - #11303 to fix those

- [ ] Cherry pick flake8 hotfix

File: src/tests/ftest/pool/create_all_vm.py (hunk 7 above)
Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.py:144]: - [ ] Fix typo

**Review by shimizukko — DISMISSED** [review]

**Comment by knard-intel** [comment]
@mjmac , please could you tell me if the latest modifications are fine with you ?
Thanks in advance, Cedric.

**Review by mjmac — DISMISSED** [review]
LGTM

File: src/tests/ftest/pool/create_all_vm.py [hunk 8]
```diff
@@ -22,7 +22,14 @@ def setUp(self):
         self.scm_avail_bytes = self.get_available_bytes()
 
     def get_available_bytes(self, ranks=None):
-        """Return the available size of SCM storage."""
+        """Return the available size of SCM storage.
+
+        Args:
+            ranks (list): List of server ranks to filter. Defaults to None.
```
Comment (ashleypittman) [inline src/tests/ftest/pool/create_all_vm.py:28]: From reading the code we typically do the following for optional args which this appears to be.  Not worth re-pushing for however.
```suggestion
            ranks (list, optional): List of server ranks to filter. Defaults to None.
```

Comment (knard-intel) [reply src/tests/ftest/pool/create_all_vm.py:28]: > From reading the code we typically do the following for optional args which this appears to be. Not worth re-pushing for however.

No problem to push it again :)

- [ ] Fix docstring

**Comment by knard-intel** [comment]
Pylint errors are unrelated to this PR

**Review by shimizukko — APPROVED** [review]

**Comment by knard-intel** [comment]
pylint errors not related to this PR
